### ISSUE TITLE
build without the bytes: "dangling symbolic link" error for symlink output

### ISSUE THREAD

**Issue by william-smith-skydio** [bazelbuild/bazel#19143]
### Description of the bug:

I have an action with an output that is a symlink to another output of the same action. When build without the bytes is enabled (e.g. `--remote_download_toplevel` or `--remote_download_minimal`), the symlink gets downloaded, but the target does not, resulting in bazel failing due to a "dangling symbolic link".

### What category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

This is a simple action which produces two outputs, one of which is a symlink to the other:

```bazel
genrule(
    name = "foobar",
    outs = [
        "a",
        "b",
    ],
    cmd = "touch $(location a) && ln -s a $(location b)",
)
```

This succeeds and populates the disk cache:

```bash
bazel build //:foobar --disk_cache /tmp/cache --remote_download_toplevel
```

Then this fails:

```bash
bazel clean
bazel build //:foobar --disk_cache /tmp/cache --remote_download_toplevel
```

Error:
```
ERROR: /<redacted>/BUILD.bazel:1:8: declared output 'b' is a dangling symbolic link
ERROR: /<redacted>/BUILD.bazel:1:8: Executing genrule //:foobar failed: not all outputs were created or valid
```

The output directory contains just a broken `b` symlink, no `a` in sight.

### Which operating system are you running Bazel on?

Ubuntu 18.04

### What is the output of `bazel info release`?

6.3.1

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

This reproduces in bazel 6.3.0 and 6.3.1, but not in 6.2.1 or 7.0.0-pre.20230710.5.

### Any other information, logs, or outputs that you want to share?

I have shown an example with a genrule and disk cache, but this also reproduces with a rule with declared outputs and a proper remote cache setup.

**Comment by brentleyjones** [comment]
@coeuvre Would a localized fix be possible? The report says that it only broke in 6.3.x, but 6.2.1 worked, so this is minor release regression?

**Comment by tjgq** [comment]
@brentleyjones I will take a look to see if a simple fix is possible.

**Comment by joeljeske** [comment]
6.4 is slated to go out soon. Is there any update on the feasibility of fixing this 6.x regression?

**Comment by tjgq** [comment]
A fix has been submitted into the 6.4.0 release branch.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.4.0 RC2](https://releases.bazel.build/6.4.0/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
